File: Java/src/main/java/chapter01/CreateStatementData.java
package chapter01;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@AllArgsConstructor
class CreateStatementData {
    Map<String, Play> plays;

    @Data
    static class Invoice {
        String customer;
        List<Performance> performances;
    }

    @Data
    @NoArgsConstructor
    static class Performance {
        public Performance(String playID, int audience) {
            this.playID = playID;
            this.audience = audience;
        }

        String playID;
        int audience;
        Play play;
        int amount;
        int volumeCredits;
    }

    @Data
    static class Play {
        String name;
        String type;
    }

    @Data
    static class StatementData {
        String customer;
        List<Performance> performances;
        int totalAmount;
        int totalVolumeCredits;
    }

    static class TragedyCalculator extends PerformanceCalculator {
        public TragedyCalculator(Performance aPerformance, Play aPlay) {
            super(aPerformance, aPlay);
        }

        @Override
        int amount() throws Error {
            int result = 40000;
            if (performance.audience > 30) {
                result += 1000 * (performance.audience - 30);
            }

            return result;
        }
    }

    static class ComedyCalculator extends PerformanceCalculator {
        public ComedyCalculator(Performance aPerformance, Play aPlay) {
            super(aPerformance, aPlay);
        }

        @Override
        int amount() throws Error {
            int result = 30000;
            if (performance.audience > 20) {
                result += 10000 + 500 * (performance.audience - 20);
            }
            result += 300 * performance.audience;

            return result;
        }
    }

    static abstract class PerformanceCalculator {
        Performance performance;
        Play play;

        public PerformanceCalculator(Performance aPerformance, Play aPlay) {
            this.performance = aPerformance;
            this.play = aPlay;
        }

        static PerformanceCalculator createPerformanceCalculator(Performance aPerformance, Play aPlay) {
            switch (aPlay.type) {
                case "tragedy":
                    return new TragedyCalculator(aPerformance, aPlay);
                case "comedy":
                    return new ComedyCalculator(aPerformance, aPlay);
                default:
                    throw new Error("unknown type: " + aPlay.type);
            }
        }

        int amount() throws Error {
            throw new Error("subclass responsibility");
        }

        int volumeCredits() {
            int result = 0;
            result += Math.max(performance.audience - 30, 0);
            if ("comedy".equals(play.type))
                result += Math.floor(performance.audience / 5);

            return result;
        }
    }

    public StatementData createStatementData(Invoice invoice, Map<String, Play> plays) throws Error {
        StatementData data = new StatementData();
        data.customer = invoice.customer;
        data.performances = invoice.performances.stream().map(this::enrichPerformance).collect(Collectors.toList());
        data.totalAmount = totalAmount(data);
        data.totalVolumeCredits = totalVolumeCredits(data);

        return data;
    }

    private Performance enrichPerformance(Performance aPerformance) {
        PerformanceCalculator calculator = PerformanceCalculator.createPerformanceCalculator(aPerformance,
                playFor(aPerformance));
        Performance result = new Performance(aPerformance.playID, aPerformance.audience);
        result.play = calculator.play;
        result.amount = calculator.amount();
        result.volumeCredits = calculator.volumeCredits();

        return result;
    };

    private int totalAmount(StatementData data) throws Error {
        return data.performances.stream().mapToInt(Performance::getAmount).sum();
    }

    private int totalVolumeCredits(StatementData data) {
        return data.performances.stream().mapToInt(Performance::getVolumeCredits).sum();
    }

    private Play playFor(Performance aPerformance) {
        return plays.get(aPerformance.playID);
    }
}

File: Rust/chapter01/src/create_statement_data.rs
use std::collections::HashMap;

pub struct Invoice<'a> {
    pub performances: Vec<Performance<'a>>,
    pub customer: &'a str,
}

pub struct Play<'a> {
    pub name: &'a str,
    pub _type: &'a str,
}

#[derive(Clone)]
pub struct Performance<'a> {
    // original fields
    pub play_id: &'a str,

    // enriched fields
    pub play: Option<&'a Play<'a>>,
    pub audience: i32,
    pub amount: Option<i32>,
    pub volume_credits: Option<i32>,
}

pub trait PerformanceCalculator<'a> {
    fn performance(&self) -> &'a Performance<'a>;
    fn play(&self) -> &'a Play<'a>;
    fn amount(&self) -> i32 {
        unimplemented!("subclass responsibility")
    }
    fn volume_credits(&self) -> i32 {
        self.super_volume_credits()
    }
    /// Only for trait default impl
    fn super_volume_credits(&self) -> i32 {
        (self.performance().audience - 30).max(0)
    }
}

/// factory method to create a performance calculator with the concrete type
fn create_performance_calculator<'a>(
    performance: &'a Performance<'a>,
    play: &'a Play<'a>,
) -> Box<dyn PerformanceCalculator<'a> + 'a> {
    match play._type {
        "tragedy" => Box::new(TragedyCalculator { performance, play }),
        "comedy" => Box::new(ComedyCalculator { performance, play }),
        _ => panic!("unknown type: {}", play._type),
    }
}

struct TragedyCalculator<'a> {
    performance: &'a Performance<'a>,
    play: &'a Play<'a>,
}

impl<'a> PerformanceCalculator<'a> for TragedyCalculator<'a> {
    fn performance(&self) -> &'a Performance<'a> {
        self.performance
    }

    fn play(&self) -> &'a Play<'a> {
        self.play
    }

    fn amount(&self) -> i32 {
        let mut result = 40000;
        if self.performance.audience > 30 {
            result += 1000 * (self.performance.audience - 30);
        }

        result
    }
}

struct ComedyCalculator<'a> {
    performance: &'a Performance<'a>,
    play: &'a Play<'a>,
}

impl<'a> PerformanceCalculator<'a> for ComedyCalculator<'a> {
    fn performance(&self) -> &'a Performance<'a> {
        self.performance
    }
    fn play(&self) -> &'a Play<'a> {
        self.play
    }
    fn amount(&self) -> i32 {
        let mut result = 30000;
        if self.performance.audience > 20 {
            result += 10000 + 500 * (self.performance.audience - 20);
        }
        result += 300 * self.performance.audience;

        result
    }
    fn volume_credits(&self) -> i32 {
        self.super_volume_credits() + (self.performance.audience / 5)
    }
}

pub struct StatementData<'a> {
    pub customer: &'a str,
    pub performances: Vec<Performance<'a>>,
    pub total_amount: Option<i32>,
    pub total_volume_credits: Option<i32>,
}

pub fn create_statement_data<'a>(
    invoice: &'a Invoice<'a>,
    plays: &'a HashMap<&'a str, Play<'a>>,
) -> StatementData<'a> {
    let play_for = |a_performance: &'a Performance<'_>| -> &'a Play<'a> {
        plays.get(a_performance.play_id).unwrap()
    };

    let enrich_performance = |a_performance: &'a Performance<'a>| {
        let calculator = create_performance_calculator(a_performance, play_for(a_performance));

        let mut mut_performance = a_performance.clone(); // Cow
        mut_performance.play = Some(calculator.play());
        mut_performance.amount = Some(calculator.amount());
        mut_performance.volume_credits = Some(calculator.volume_credits());

        mut_performance
    };

    let total_amount = |data: &StatementData| -> i32 {
        data.performances
            .iter()
            .fold(0, |acc, perf| acc + perf.amount.unwrap())
    };

    let total_volume_credits = |data: &StatementData| -> i32 {
        data.performances
            .iter()
            .fold(0, |acc, perf| acc + perf.volume_credits.unwrap())
    };

    let performances = invoice
        .performances
        .iter()
        .map(enrich_performance)
        .collect();

    let mut statement_data = StatementData {
        customer: invoice.customer,
        performances,
        total_amount: None,
        total_volume_credits: None,
    };

    statement_data.total_amount = Some(total_amount(&statement_data));
    statement_data.total_volume_credits = Some(total_volume_credits(&statement_data));
    statement_data
}
